mitiq/executor/executor.py

```diff
@@ -160,38 +185,55 @@ def evaluate(
                 for circuit_with_measurements in observable.measure_in(circuit)
             ]
             result_step = observable.ngroups
-        elif (
-            observable is not None
-            and self._executor_return_type not in MeasurementResultLike
-            and self._executor_return_type not in DensityMatrixLike
-        ):
-            raise ValueError(
-                """Executor and observable are not compatible. Executors
-                returning expectation values as float must be used with
-                observable=None"""
-            )
         else:
             all_circuits = circuits
             result_step = 1
 
         # Run all required circuits.
-        all_results = self.run(all_circuits, force_run_all, **kwargs)
+        try:
+            all_results = self.run(all_circuits, force_run_all, **kwargs)
+        except Exception:
+            if observable is not None and self._executor_return_type is None:
+                all_circuits = [
+                    circuit_with_measurements
+                    for circuit in circuits
+                    for circuit_with_measurements in observable.measure_in(
+                        circuit
+                    )
+                ]
+                all_results = self.run(all_circuits, force_run_all, **kwargs)
+            else:
+                raise
+
+        # check returned type
+        manual_return_type = None
+        if len(all_results) > 0:
+            manual_return_type = type(all_results[0])
```

Comment:
Should we even care about the `_executor_return_type` at this point?

Reply:
If an executor returns something different than what is specified there will be other errors thrown. You are right, at this point we can stick with the `manual_return_type`.

Reply:
Addressed in https://github.com/unitaryfund/mitiq/pull/2514/commits/3c7b02737fad652628d740b23633fac6fcb78607

Reply:
Should I add more unit testing since I just introduced this `if` branching?